File: frontend/web/services/workflows/editor.ts
import { isAxiosError } from "axios";
import { EditorWorkflow } from "@/types/workflows";
import { ServiceReturn } from "@/types/api";
import { axiosInstance } from "@/services";
import { SERVICE_ERROR_UNKNOWN } from "@/config/services";
import { convertAreaParamsToWorkflowPayloadParams } from "@/utils/workflows";

type CreateWorkflowReturn = {
	id: string;
};
export const create = async (workflow: EditorWorkflow): Promise<ServiceReturn<CreateWorkflowReturn>> => {
	try {
		const response = await axiosInstance.post<CreateWorkflowReturn>(`/workflows`, {
			name: workflow.name,
			active: workflow.active,
			action: {
				id: workflow.action.id,
				parameters: convertAreaParamsToWorkflowPayloadParams(workflow.action.area!.parameters),
				areaId: workflow.action.area?.id,
				areaServiceId: workflow.action.areaService?.id,
			},
			reactions: workflow.reactions.map((reaction) => ({
				id: reaction.id,
				parameters: convertAreaParamsToWorkflowPayloadParams(reaction.area!.parameters),
				areaId: reaction.area?.id,
				areaServiceId: reaction.areaService?.id,
				previousAreaId: reaction.previousAreaId,
			})),
		});
		return { data: response.data, error: undefined };
	} catch (error) {
		if (isAxiosError(error)) {
			return { data: null, error: error.response?.data?.message ?? SERVICE_ERROR_UNKNOWN };
		}
		return { data: null, error: SERVICE_ERROR_UNKNOWN };
	}
};

export const update = async (
	workflow: Partial<EditorWorkflow> & {
		id: string;
	},
): Promise<ServiceReturn<CreateWorkflowReturn>> => {
	try {
		const { id } = workflow;
		const response = await axiosInstance.patch<CreateWorkflowReturn>(`/workflows/${id}`, {
			name: workflow.name,
			action: {
				id: workflow.action!.id,
				parameters: convertAreaParamsToWorkflowPayloadParams(workflow.action!.area!.parameters),
				areaId: workflow.action!.area?.id,
				areaServiceId: workflow.action!.areaService?.id,
			},
			reactions: workflow.reactions!.map((reaction) => ({
				id: reaction.id,
				parameters: convertAreaParamsToWorkflowPayloadParams(reaction.area!.parameters),
				areaId: reaction.area?.id,
				areaServiceId: reaction.areaService?.id,
				previousAreaId: reaction.previousAreaId,
			})),
		});
		return { data: response.data, error: undefined };
	} catch (error) {
		return { data: null, error: SERVICE_ERROR_UNKNOWN };
	}
};

export const rename = async (id: string, name: string) => {
	try {
		const response = await axiosInstance.patch<CreateWorkflowReturn>(`/workflows/${id}`, {
			name,
		});
		return { data: response.data, error: undefined };
	} catch (error) {
		return { data: null, error: SERVICE_ERROR_UNKNOWN };
	}
};

File: frontend/mobile/lib/services/workflows/editor.dart
import 'package:area_mobile/services/dio.dart';
import 'package:area_mobile/types/services.dart';
import 'package:area_mobile/types/workflows/editor.dart';
import 'package:area_mobile/utils/workflows.dart';

Future<ServiceReturn<int>> create(EditorWorkflow workflow) async {
  try {
    final response = await dio.post('/workflows', data: {
      "name": workflow.name,
      "active": workflow.active,
      "action": {
        "id": workflow.action.id,
        "parameters": convertAreaParamsToWorkflowPayloadParams(
            workflow.action.area!.parameters),
        "areaId": workflow.action.area?.id,
        "areaServiceId": workflow.action.areaService?.id,
      },
      "reactions": workflow.reactions
          .map((reaction) => ({
                "id": reaction.id,
                "parameters": convertAreaParamsToWorkflowPayloadParams(
                    reaction.area!.parameters),
                "areaId": reaction.area?.id,
                "areaServiceId": reaction.areaService?.id,
                "previousAreaId": reaction.previousAreaId,
              }))
          .toList(),
    });
    return ServiceReturn(data: response.statusCode);
  } catch (e) {
    return ServiceReturn(error: e.toString());
  }
}

Future<ServiceReturn<int>> update(EditorWorkflow workflow) async {
  try {
    final response = await dio.patch('/workflows/${workflow.id}', data: {
      "name": workflow.name,
      "action": {
        "id": workflow.action.id,
        "parameters": convertAreaParamsToWorkflowPayloadParams(
            workflow.action.area!.parameters),
        "areaId": workflow.action.area?.id,
        "areaServiceId": workflow.action.areaService?.id,
      },
      "reactions": workflow.reactions
          .map((reaction) => ({
                "id": reaction.id,
                "parameters": convertAreaParamsToWorkflowPayloadParams(
                    reaction.area!.parameters),
                "areaId": reaction.area?.id,
                "areaServiceId": reaction.areaService?.id,
                "previousAreaId": reaction.previousAreaId,
              }))
          .toList(),
    });
    return ServiceReturn(data: response.statusCode);
  } catch (e) {
    return ServiceReturn(error: e.toString());
  }
}
